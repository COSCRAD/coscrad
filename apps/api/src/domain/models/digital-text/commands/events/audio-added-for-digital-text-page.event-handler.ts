import { Inject } from '@nestjs/common';
import { EventHandler, ICoscradEventHandler } from '../../../../../domain/common';
import { isInternalError } from '../../../../../lib/errors/InternalError';
import { isNotFound } from '../../../../../lib/types/not-found';
import { DigitalTextViewModel } from '../../../../../queries/digital-text';
import { IAggregateRootQueryRepository } from '../../../../../queries/interfaces';
import { AudioAddedForDigitalTextPage } from './audio-added-for-digital-text-page.event';

@EventHandler(`AUDIO_ADDED_FOR_DIGITAL_TEXT_PAGE`)
export class AudioAddedForDigitalTextPageEventHandler implements ICoscradEventHandler {
    constructor(
        @Inject(`DIGITAL_TEXT_QUERY_REPOSITORY`)
        private readonly digitalTextQueryRepository: IAggregateRootQueryRepository<DigitalTextViewModel>
    ) {}

    async handle({
        payload: {
            aggregateCompositeIdentifier: { id },
            audioItemId,
            languageCode,
        },
    }: AudioAddedForDigitalTextPage): Promise<void> {
        /**
         * It's possible that we can actually just do the update without ever
         * fetching. At some point, we should measure the performance of this.
         * For example, we could have
         * this.digitalTextRepository.insertItemIntoArray('audio.items',{
         *     languageCode,
         *     audioItemId
         * })
         */
        const existingDigitalText = await this.digitalTextQueryRepository.fetchById(id);

        // Once we use a messaging queue, we can fail with an exception. For now, we don't want to interrupt the synchronous thread.
        if (isNotFound(existingDigitalText)) return;

        /**
         * TODO Shouldn't we keep a denormalized view of the audio item here?
         * What does that look like?
         */
        const updatedAudio = existingDigitalText.audio.addAudio(audioItemId, languageCode);

        // ignore invalid updates to audio
        if (isInternalError(updatedAudio)) return;

        // We should append the applied event as well
        await this.digitalTextQueryRepository.update({
            id,
            audio: updatedAudio,
        });
    }
}
