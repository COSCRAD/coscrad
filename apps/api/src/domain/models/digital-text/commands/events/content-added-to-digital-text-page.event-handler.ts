import { Inject } from '@nestjs/common';
import { EventHandler, ICoscradEventHandler } from '../../../../../domain/common';
import { isNotFound } from '../../../../../lib/types/not-found';
import { DigitalTextViewModel } from '../../../../../queries/digital-text';
import { IAggregateRootQueryRepository } from '../../../../../queries/interfaces';
import { ContentAddedToDigitalTextPage } from './content-added-to-digital-text-page.event';

@EventHandler(`CONTENT_ADDED_TO_DIGITAL_TEXT_PAGE`)
export class ContentAddedToDigitalTextPageEventHandler implements ICoscradEventHandler {
    constructor(
        @Inject(`DIGITAL_TEXT_QUERY_REPOSITORY`)
        private readonly digitalTextQueryRepository: IAggregateRootQueryRepository<DigitalTextViewModel>
    ) {}

    async handle(event: ContentAddedToDigitalTextPage): Promise<void> {
        const {
            payload: {
                aggregateCompositeIdentifier: { id },
            },
        } = event;

        const existingDigitalText = await this.digitalTextQueryRepository.fetchById(id);

        // TODO Log this occurrence
        if (isNotFound(existingDigitalText)) return;

        // TODO Is this really the pattern we want?
        const updatedDigitalText = existingDigitalText.apply(event);

        await this.digitalTextQueryRepository.update(updatedDigitalText);
    }
}
