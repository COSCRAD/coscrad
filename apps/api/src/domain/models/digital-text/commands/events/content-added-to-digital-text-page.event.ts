import { CoscradEvent } from '../../../../common';
import { BaseEvent } from '../../../shared/events/base-event.entity';
import { AddContentToDigitalTextPage } from '../add-content-to-digital-text-page/add-content-to-digital-text-page.command';

export type ContentAddedToDigitalTextPagePayload = AddContentToDigitalTextPage;

const eventType = 'CONTENT_ADDED_TO_DIGITAL_TEXT_PAGE';

/**
 * TODO It seems that if event type is a constant we might be able to use reflection
 * to access this value on the prototype chain and avoid needing an argument
 * in the following decorator factory.
 */
@CoscradEvent(eventType)
export class ContentAddedToDigitalTextPage extends BaseEvent<ContentAddedToDigitalTextPagePayload> {
    type = eventType;
}
