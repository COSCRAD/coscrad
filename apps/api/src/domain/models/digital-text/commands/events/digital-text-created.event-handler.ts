import { Inject } from '@nestjs/common';
import { EventHandler, ICoscradEventHandler } from '../../../../../domain/common';
import { buildMultilingualTextWithSingleItem } from '../../../../../domain/common/build-multilingual-text-with-single-item';
import { DigitalTextViewModel } from '../../../../../queries/digital-text';
import { IAggregateRootQueryRepository } from '../../../../../queries/interfaces';
import { MultilingualAudio } from '../../../shared/multilingual-audio/multilingual-audio.entity';
import { DigitalTextCreated } from './digital-text-created.event';

@EventHandler(`DIGITAL_TEXT_CREATED`)
export class DigitalTextCreatedEventHandler implements ICoscradEventHandler {
    constructor(
        @Inject(`DIGITAL_TEXT_QUERY_REPOSITORY`)
        private readonly digitalTextQueryRepository: IAggregateRootQueryRepository<DigitalTextViewModel>
    ) {}

    async handle(event: DigitalTextCreated): Promise<void> {
        const {
            payload: {
                aggregateCompositeIdentifier: { id },
                title: titleText,
                languageCodeForTitle,
            },
        } = event;

        const title = buildMultilingualTextWithSingleItem(titleText, languageCodeForTitle);

        const newDigitalText = DigitalTextViewModel.fromSnapshot({
            id,
            name: title,
            title: title,
            pages: [],
            tags: [],
            isPublished: false,
            audio: new MultilingualAudio({
                items: [],
            }),
        });

        await this.digitalTextQueryRepository.create(newDigitalText);
    }
}
