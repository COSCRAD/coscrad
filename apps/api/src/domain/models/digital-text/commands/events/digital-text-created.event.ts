import { AggregateType, LanguageCode, MultilingualTextItemRole } from '@coscrad/api-interfaces';
import { NestedDataType } from '@coscrad/data-types';
import { MultilingualTextItem } from '../../../../../domain/common/entities/multilingual-text';
import { AggregateCompositeIdentifier } from '../../../../../domain/types/AggregateCompositeIdentifier';
import { CoscradEvent } from '../../../../common';
import { BaseEvent } from '../../../shared/events/base-event.entity';
import { EventRecordMetadata } from '../../../shared/events/types/EventRecordMetadata';
import { DIGITAL_TEXT_CREATED } from '../../constants';
import {
    CreateDigitalText,
    DigitalTextCompositeId,
} from '../create-digital-text/create-digital-text.command';

export class DigitalTextCreatedPayload {
    @NestedDataType(DigitalTextCompositeId, {
        label: 'Composite Identifier',
        description: 'system-wide unique identifier',
    })
    readonly aggregateCompositeIdentifier: AggregateCompositeIdentifier<
        typeof AggregateType.digitalText
    >;

    @NestedDataType(MultilingualTextItem, {
        label: 'title',
        description: `multilingual text item definition for the digital text's title`,
    })
    readonly title: MultilingualTextItem;

    constructor({
        titleText,
        languageCodeForTitle,
        aggregateCompositeIdentifier,
    }: {
        titleText: string;
        languageCodeForTitle: LanguageCode;
        aggregateCompositeIdentifier: AggregateCompositeIdentifier<
            typeof AggregateType.digitalText
        >;
    }) {
        this.aggregateCompositeIdentifier = aggregateCompositeIdentifier;

        this.title = new MultilingualTextItem({
            text: titleText,
            languageCode: languageCodeForTitle,
            // The event payload encodes the significance of this translation for consumers
            role: MultilingualTextItemRole.original,
        });
    }
}

@CoscradEvent(DIGITAL_TEXT_CREATED)
export class DigitalTextCreated extends BaseEvent<DigitalTextCreatedPayload> {
    type = DIGITAL_TEXT_CREATED;

    @NestedDataType(DigitalTextCreatedPayload, {
        label: 'payload',
        description: 'the event payload',
    })
    payload: DigitalTextCreatedPayload;

    constructor(
        { title, languageCodeForTitle, aggregateCompositeIdentifier }: CreateDigitalText,
        meta: EventRecordMetadata
    ) {
        super(
            new DigitalTextCreatedPayload({
                aggregateCompositeIdentifier,
                titleText: title,
                languageCodeForTitle,
            }),
            meta
        );
    }
}
