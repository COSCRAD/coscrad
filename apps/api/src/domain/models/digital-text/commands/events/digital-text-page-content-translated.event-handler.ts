import { Inject } from '@nestjs/common';
import { isInternalError } from '../../../../../lib/errors/InternalError';
import { isNotFound } from '../../../../../lib/types/not-found';
import { DigitalTextViewModel } from '../../../../../queries/digital-text';
import { IAggregateRootQueryRepository } from '../../../../../queries/interfaces';
import { DigitalTextPageContentTranslated } from './digital-text-page-content-translated.event';

export class DigitalTextPageContentTranslatedEventHandler {
    constructor(
        @Inject(`DIGITAL_TEXT_QUERY_REPOSITORY`)
        private readonly digitalTextQueryRepository: IAggregateRootQueryRepository<DigitalTextViewModel>
    ) {}

    async handle(event: DigitalTextPageContentTranslated): Promise<void> {
        const {
            payload: {
                aggregateCompositeIdentifier: { id },
                pageIdentifier,
                translation,
                languageCode,
            },
        } = event;

        const existingDigitalText = await this.digitalTextQueryRepository.fetchById(id);

        // TODO Log
        if (isNotFound(existingDigitalText)) return;

        const targetPage = existingDigitalText.pages.find(
            (page) => page.identifier === pageIdentifier
        );

        if (!targetPage) return;

        const pageUpdateResult = targetPage.translateContent(translation, languageCode);

        if (isInternalError(pageUpdateResult)) {
            return;
        }

        const updatedPages = existingDigitalText.pages.map((page) =>
            page.identifier === id ? pageUpdateResult : page
        );

        this.digitalTextQueryRepository.update({
            id,
            pages: updatedPages,
        });
    }
}
