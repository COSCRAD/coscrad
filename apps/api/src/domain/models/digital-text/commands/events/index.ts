export * from './audio-added-for-digital-text-page.event';
export * from './content-added-to-digital-text-page.event';
export * from './digital-text-created.event';
export * from './page-added-to-digital-text.event';
