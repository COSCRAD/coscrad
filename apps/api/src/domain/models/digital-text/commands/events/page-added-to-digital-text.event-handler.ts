import { Inject } from '@nestjs/common';
import { EventHandler, ICoscradEventHandler } from '../../../../../domain/common';
import { isNotFound } from '../../../../../lib/types/not-found';
import { DigitalTextViewModel } from '../../../../../queries/digital-text';
import { IAggregateRootQueryRepository } from '../../../../../queries/interfaces';
import { MultilingualAudio } from '../../../shared/multilingual-audio/multilingual-audio.entity';
import DigitalTextPage from '../../entities/digital-text-page.entity';
import { PageAddedToDigitalText } from './page-added-to-digital-text.event';

@EventHandler(`PAGE_ADDED_TO_DIGITAL_TEXT`)
export class PageAddedToDigitalTextEventHandler implements ICoscradEventHandler {
    constructor(
        @Inject(`DIGITAL_TEXT_QUERY_REPOSITORY`)
        private readonly digitalTextQueryRepository: IAggregateRootQueryRepository<DigitalTextViewModel>
    ) {}

    async handle(event: PageAddedToDigitalText): Promise<void> {
        const {
            payload: {
                aggregateCompositeIdentifier: { id },
                identifier,
            },
        } = event;

        const existingDigitalText = await this.digitalTextQueryRepository.fetchById(id);

        // TODO Log
        if (isNotFound(existingDigitalText)) return;

        await this.digitalTextQueryRepository.update({
            id,
            pages: existingDigitalText.pages.concat(
                new DigitalTextPage({
                    identifier,
                    audio: new MultilingualAudio({
                        items: [],
                    }),
                })
            ),
        });
    }
}
