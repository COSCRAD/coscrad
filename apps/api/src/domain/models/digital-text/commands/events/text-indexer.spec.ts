import { LanguageCode } from '@coscrad/api-interfaces';
import { Test } from '@nestjs/testing';
import { TestEventStream } from '../../../../../test-data/events/test-event-stream';
import buildDummyUuid from '../../../__tests__/utilities/buildDummyUuid';
import { DigitalTextCreated } from './digital-text-created.event';
import {
    CoscradIndex,
    ITextIndexer,
    ITokenizer,
    TextIndexingEventConsumer,
} from './text-indexer.event-handler';

class InMemoryTextIndexer implements ITextIndexer {
    private readonly indexDocuments: CoscradIndex[] = [];

    index(...indexes: CoscradIndex[]): Promise<void> {
        indexes.forEach((index) => this.indexSingle(index));

        return Promise.resolve();
    }

    // this is for POC only. We would probably only expose `search` functionality and then test this from the outside.
    public fetchIndexes(): CoscradIndex[] {
        return this.indexDocuments;
    }

    private indexSingle(index: CoscradIndex) {
        // TODO In reality, we'd need to merge if the token is already known this is just a POC
        this.indexDocuments.push(index);
    }
}

class InMemoryTokenizer implements ITokenizer {
    tokenize(longText: string): string[] {
        return longText
            .split(' ')
            .flatMap(
                // cartoonish only
                (withPunctuation) => withPunctuation.split('.')
            )
            .map((caseDependent) => caseDependent.toLowerCase());
    }
}

const mockIndexer = new InMemoryTextIndexer();

describe(`TextIndexer event consumer`, () => {
    let textIndexingEventConsumer: TextIndexingEventConsumer;

    beforeAll(async () => {
        const testModuleRef = await Test.createTestingModule({
            providers: [
                {
                    provide: 'TEXT_INDEXER_TOKEN',
                    useValue: mockIndexer,
                },
                {
                    provide: 'TOKENIZER_INJECTION_TOKEN',
                    useValue: new InMemoryTokenizer(),
                },
                {
                    provide: DigitalTextCreated,
                    useValue: DigitalTextCreated,
                },
                TextIndexingEventConsumer,
            ],
        }).compile();

        const app = testModuleRef.createNestApplication();

        await app.init();

        textIndexingEventConsumer = app.get(TextIndexingEventConsumer);
    });

    describe(`when indexing text`, () => {
        const languageCode = LanguageCode.English;

        const titleText = 'A big dog runs fast';

        const digitalTextCreated = new TestEventStream().andThen<DigitalTextCreated>({
            type: `DIGITAL_TEXT_CREATED`,
            payload: {
                title: {
                    text: titleText,
                    languageCode,
                },
            },
        });

        it.only(`should work`, async () => {
            await textIndexingEventConsumer.handle(
                digitalTextCreated.as({
                    id: buildDummyUuid(555),
                })[0]
            );

            const updatedIndexes = mockIndexer.fetchIndexes();

            expect(updatedIndexes).not.toEqual([]);

            expect(updatedIndexes).toMatchSnapshot();
        });
    });
});
