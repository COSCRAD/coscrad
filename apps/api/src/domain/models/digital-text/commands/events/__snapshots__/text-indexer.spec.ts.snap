// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextIndexer event consumer when indexing text should work 1`] = `
[
  {
    "languageCode": "en",
    "token": "title",
    "viewCompositeIdentifiers": [
      {
        "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100555",
        "type": "digitalText",
      },
    ],
  },
  {
    "languageCode": "en",
    "token": "for",
    "viewCompositeIdentifiers": [
      {
        "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100555",
        "type": "digitalText",
      },
    ],
  },
  {
    "languageCode": "en",
    "token": "digital",
    "viewCompositeIdentifiers": [
      {
        "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100555",
        "type": "digitalText",
      },
    ],
  },
  {
    "languageCode": "en",
    "token": "text",
    "viewCompositeIdentifiers": [
      {
        "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b100555",
        "type": "digitalText",
      },
    ],
  },
]
`;
